062-worker-pool/proposal.md

```diff
@@ -0,0 +1,157 @@
+# Summary
+
+This proposal outlines the support for worker pool. A worker pool is a group
+of Concourse workers, that can be associated with one or multiple teams.
+
+# Motivation
+
+One or multiple Concourse workers can be associated with a team by adding 
+`--team <team name>` option when launching `concourse worker`. Then all 
+pipeline containers of the team will be dispatched to the workers. The 
+relationship between workers and teams are N-to-1:
+
+```
+  Workers (0..n) <----------------> (0..1) Team
+```
+
+This is a useful mechanism to isolate some teams' workload or to provide
+dedicate workers to some teams.
+
+In practice, some big projects may split their pipelines to multiple small
+teams. When a project that has multiple teams requests dedicate workers,
+it is hard to give. Because the project actually wants some workers sharing
+with all of its teams, but Concourse only support workers per team.
+
+For example, a big project of my company has 100 teams. By evaluating
+workload of those teams, the project requests 20 dedicate workers. However, due
+to the restriction of 1 worker serving at most 1 team, we have to deploy at
+least 100 workers (1 worker for a team) to the project. Moreover, each team 
+having a single worker is unreliable, if the worker is down, then all builds of
+the team will fail. If we deploy 2 workers for each team for redundancy, then
+we need to deploy 200 workers. Comparing 200 workers we need to deploy with 20
+workers they request, it is a huge waste.
+
+# Proposal
+
+With introducing a new concept "worker pool", we build N-N relationship between
+workers and teams:
+
+```
+  Workers (0..n) <----> (1) Worker Pool (1) <------> (0..n) Teams
+```
+
+* A worker pool may contain 0 or multiple workers.
+* A worker can be added to at most 1 worker pool.
+* A worker may also not belong to any worker pool.
+* If a worker is added to a worker pool, then it cannot be associated with a
+team directly.
+* If a worker is not added to any worker pool, then it can still be associated
+with a team directly for backward-compatibility.
+* A worker pool may be assign to 0 or multiple teams.
+* A team can be associated with at most 1 worker pool.
+* A team can still be associated with a worker directly for backward-compatibility.
+* A team may not associate with any worker pool or team for backward-compatibility.
+
+The logic of choosing a worker for a step will look like:
+
+```go
+   team := step.TeamName()
+   var candidateWorkers = []Workers
+   if team.WorkerPool() != nil {
+      candidateWorkers = team.WorkerPool().Workers()
+   } else if team.Workers() != nil {
+      candidateWorkers = team.Workers()
+   } else {
```

Comment:
Having this be either-of seems kind of restricting - for example, what if a team wants to use the shared pool, but also has their own self-maintained worker that they would like to register? For example, sometimes teams will use a shared pool of workers but also run their own Mac Mini for running Darwin builds for iOS development.

I guess we could represent that as its own pool, but that would also imply allowing a team to have multiple pools.

Reply:
This is a good point. I agree that a team may have multiple pools.